lib/ruby_smb/dcerpc/lsarpc.rb

```diff
@@ -1,11 +1,644 @@
+      class LsaprSecurityDescriptorControl < Ndr::NdrStruct
+        default_parameters byte_align: 4
+        endian :little
+
+        ndr_uint16 :security_descriptor_control
+      end
+
+      class LsaprSecurityDescriptorControlPtr < LsaprSecurityDescriptorControl
+        extend Ndr::PointerClassPlugin
+      end
+
+      # [2.2.5 LSAPR_SECURITY_DESCRIPTOR](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-lsat/8494008f-0bfb-45b8-bb6c-e32dd7f18e3d)
+      class LsaprSecurityDescriptor < Ndr::NdrStruct
+        default_parameters byte_align: 4
+        endian :little
+
+        ndr_uint8                         :revision
+        ndr_uint8                         :sbz1
+        lsapr_security_descriptor_control :control
+        prpc_sid                          :owner
+        prpc_sid                          :group
+        lsapr_acl_ptr                     :sacl
+        lsapr_acl_ptr                     :dacl
+      end
+
+      class LsaprSecurityDescriptorPtr < LsaprSecurityDescriptor
+        extend Ndr::PointerClassPlugin
+      end
+
+      # [2.2.3.5 SECURITY_IMPERSONATION_LEVEL](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-lsad/720cea10-cee2-4c45-9084-c6fa7d67d18d)
+      class SecurityImpersonationLevel < Ndr::NdrStruct
+        default_parameters byte_align: 4
+        endian :little
+
+        ndr_uint32 :level # This is SECURITY_IMPERSONATION_LEVEL, type 'enum' -> uint32
+      end
+
+      class SecurityImpersonationLevelPtr < SecurityImpersonationLevel
+        extend Ndr::PointerClassPlugin
+      end
+
+      # [2.2.3.6 SECURITY_CONTEXT_TRACKING_MODE](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-lsad/6bb42770-b924-41ff-8a57-83e37b8b7797)
+      class LsaprSecurityContextTrackingMode < Ndr::NdrStruct
+        default_parameters byte_align: 4
+        endian :little
+
+        ndr_uint8 :security_context_tracking_mode
+      end
+
+      class LsaprSecurityContextTrackingModePtr < LsaprSecurityContextTrackingMode
+        extend Ndr::PointerClassPlugin
+      end
+
+      # [2.2.3.7 SECURITY_QUALITY_OF_SERVICE](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-lsad/0ddf3150-53b5-42a5-b0ec-518bce67738c)
+      class SecurityQualityOfService < Ndr::NdrStruct
+        default_parameters byte_align: 4
+        endian :little
+
+        ndr_uint32 :len
+        ndr_uint16 :impersonation_level
+        ndr_uint8  :security_context_tracking_mode
+        ndr_uint8  :effective_only
+      end
+
+      class SecurityQualityOfServicePtr < SecurityQualityOfService
+        extend Ndr::PointerClassPlugin
+      end
+
+      # [2.2.2.4 LSAPR_OBJECT_ATTRIBUTES](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-lsad/ad9e183d-6474-4641-a6d9-d3796d2d604b)
+      class LsaprObjectAttributes < Ndr::NdrStruct
+        default_parameters byte_align: 4
+        endian :little
+
+        ndr_uint32                      :len
```

Comment:
So I'm looking at [`#open_policy_2`](https://github.com/rapid7/metasploit-framework/pull/19170/files#diff-32c8b2752caef945fe888b76e002c9ce16113645fe0b720992b9f51ed214e988R52) and I'm wondering if it'd be possible to have this `len` field be calculated by default so the caller doesn't have to set it in most cases. Same thing goes for the `len` field of `SecurityQualityOfService`.